worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/adapter/Regenerator.java

```diff
@@ -0,0 +1,525 @@
+    private final Long2ObjectOpenHashMap<Chunk> chunks = new Long2ObjectOpenHashMap<>();
+    private ExecutorService executor;
+    private SingleThreadQueueExtent source;
+
+    /**
+     * Initializes an abstract regeneration handler.
+     * @param originalBukkitWorld the Bukkit world containing all the information on how to regenerate the {code Region}
+     * @param region the selection to regenerate
+     * @param target the target {@code Extent} to paste the regenerated blocks into
+     * @param options the options to used while regenerating and pasting into the target {@code Extent}
+     */
+    public Regenerator(org.bukkit.World originalBukkitWorld, Region region, Extent target, RegenOptions options) {
+        this.originalBukkitWorld = originalBukkitWorld;
+        this.region = region;
+        this.target = target;
+        this.options = options;
+    }
+
+    /**
+     * Regenerates the selected {@code Region}.
+     * @return whether or not the regeneration process was successful
+     * @throws Exception when something goes terribly wrong
+     */
+    public boolean regenerate() throws Exception {
+        if (!prepare()) {
+            return false;
+        }
+
+        try {
+            if (!initNewWorld()) {
+                cleanup0();
+                return false;
+            }
+        } catch (Exception e) {
+            cleanup0();
+            throw e;
+        }
+
+        try {
+            if (!generate()) {
+                cleanup0();
+                return false;
+            }
+        } catch (Exception e) {
+            cleanup0();
+            throw e;
+        }
+
+        try {
+            copyToWorld();
+        } catch (Exception e) {
+            cleanup0();
+            throw e;
+        }
+
+        cleanup0();
+        return true;
+    }
+
+    /**
+     * Returns the {@code ProtoChunk} at the given chunk coordinates.
+     * @param x the chunk x coordinate
+     * @param z the chunk z coordinate
+     * @return the {@code ProtoChunk} at the given chunk coordinates or null if it is not part of the regeneration process or has not been initialized yet.
+     */
+    protected ProtoChunk getProtoChunkAt(int x, int z) {
+        return protoChunks.get(MathMan.pairInt(x, z));
+    }
+
+    /**
+     * Returns the {@code Chunk} at the given chunk coordinates.
+     * @param x the chunk x coordinate
+     * @param z the chunk z coordinate
+     * @return the {@code Chunk} at the given chunk coordinates or null if it is not part of the regeneration process or has not been converted yet.
+     */
+    protected Chunk getChunkAt(int x, int z) {
+        return chunks.get(MathMan.pairInt(x, z));
+    }
+
+    private boolean generate() throws Exception {
+        if (generateConcurrent) {
+            //Using concurrent chunk generation
+            executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
```

Comment:
Forgot about that before, but this should either use `Settings.IMP.QUEUE.PARALLEL_THREADS` or have a separate config option. Using all cores might have a negative impact on other processes (other minecraft instances, teamspeak servers etc). @IntellectualSites/fawe-team thoughts?

Reply:
it should use the parallel threads or the actual FAWE async executor which is limited to the parallel threads defined.

Reply:
Settings.IMP.QUEUE.PARALLEL_THREADS does make sense
i'll change it

Reply:
done